### PR TITLE
add eagerCancel to Partition and Balance stage

### PR DESCRIPTION
**Pull request by Hawstein**
Ref #22234

### CODE REVIEW

**Comment by hepin1989** [comment]
Just an idea,how about make it a configurable int value instead of just a boolean? then you could simulate the current boolean one on top of that.

**Comment by Hawstein** [comment]
@hepin1989 Hmm, I dont see much value of this flexibility. Let's see others' opinion.

**Comment by johanandren** [comment]
I know that we have gone with a parameter for this in other stages, but wouldn't this be a good thing to use attributes for instead? That way there would be explosion in factory method parameters over time etc.

**Comment by johanandren** [comment]
Another idea: what if this was achieved through composition instead of adding it to multiple stages, a "synced cancelation" stage which can be put in many branches and will cancel all as soon as one of them cancels?

File: akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala [hunk 1]
```diff
@@ -164,6 +164,16 @@ object Partition {
    *
    * @param outputCount number of output ports
    * @param partitioner function deciding which output each element will be targeted
+   * @param eagerCancel if true, partition cancels upstream if any of its downstreams cancel.
+   */
+  def create[T](outputCount: Int, partitioner: function.Function[T, Int], eagerCancel: Boolean): Graph[UniformFanOutShape[T, T], NotUsed] =
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala]: Simple paramaters before lambda would probably be nicer from the users point of view

**Comment by raboof** [comment]
PLS BUILD

**Comment by MihaiAnei** [comment]
Hey guys. I tried to build a Graph using Partition and one of the partitions` outlet was returned as a SourceShape in order to be used in another Flow, which, if is Canceled, the whole Partition should Cancel (my bad on this part, I should have read the documentation more thoroughly). 
I implemented this whole behavior using Broadcast instead, and additional filtering on every outlet of every partition.
I hope my example will be motivational enough for this PR to happen :)

Then I found this pull request and I realized that every "fan-out" stage in a stream, maybe, should offer the possibility of choosing when to cancel: when all downstreams "blow up" or when at least one.

I'm looking forward for this PR. Cheers!

**Comment by hepin1989** [comment]
@Hawstein Ping~~

File: akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala [hunk 2]
```diff
@@ -147,7 +147,7 @@ object Broadcast {
  * '''Completes when''' upstream completes
  *
  * '''Cancels when'''
- *   when one of the downstreams cancel
+ *   If eagerCancel is enabled: when any downstream cancels; otherwise: when all downstreams cancel
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala]: There's probably a mention in the `stage-overview.md` as well.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 3]
```diff
@@ -480,8 +480,9 @@ object Partition {
    *
    * @param outputPorts number of output ports
    * @param partitioner function deciding which output each element will be targeted
+   * @param eagerCancel if true, partition cancels upstream if any of its downstreams cancel
    */
-  def apply[T](outputPorts: Int, partitioner: T ⇒ Int): Partition[T] = new Partition(outputPorts, partitioner)
+  def apply[T](outputPorts: Int, partitioner: T ⇒ Int, eagerCancel: Boolean = false): Partition[T] = new Partition(outputPorts, partitioner, eagerCancel)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: In general default params are a real pain in public APIs as they making evolving without breaking bincomp hard, please provide overloads instead.

**Review by johanandren — COMMENTED** [review]
Some comments added. Also missing test coverage for the change, please add.

**Comment by johanandren** [comment]
I'm afraid we have changed how the MiMa filters are added since this PR was created, please move those to akka-stream/src/main/mima-filters/2.5.10.backwards.excludes

**Comment by hepin1989** [comment]
@Hawstein Only `Balance` left.

**Comment by Hawstein** [comment]
@hepin1989 @johanandren @raboof  PR is updated.

Since Klang has already finished the modification of Partition, I delete that part. Also, a unit test is added.

Thanks for pinging me @hepin1989  ;-)

**Comment by hepin1989** [comment]
@ktoso Will this be in 2.5.12?

**Comment by ktoso** [comment]
Maybe, needs reviews

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 4]
```diff
@@ -848,11 +851,15 @@ object Balance {
...
  *
- * '''Cancels when''' all downstreams cancel
+ * '''Cancels when''' If eagerCancel is enabled: when any downstream cancels; otherwise: when all downstreams cancel
  */
-final class Balance[T](val outputPorts: Int, val waitForAllDownstreams: Boolean) extends GraphStage[UniformFanOutShape[T, T]] {
+final class Balance[T](val outputPorts: Int, val waitForAllDownstreams: Boolean, val eagerCancel: Boolean) extends GraphStage[UniformFanOutShape[T, T]] {
   // one output might seem counter intuitive but saves us from special handling in other places
   require(outputPorts >= 1, "A Balance must have one or more output ports")
+
+  @deprecated("Use the constructor which also specifies the `eagerCancel` parameter")
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: this should include the version number, so 2.5.12 if we can get it in before we release that

**Review by johanandren — APPROVED** [review]
LGTM with the version added to the deprecated annotation.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 4 above)
Comment (hepin1989) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: @Hawstein ,with Java's `@Deprecated ` and with `since 2.5.12` info.

Comment (Hawstein) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Done. @johanandren @hepin1989

**Comment by Hawstein** [comment]
@ktoso Needs another approval. Please take a look if you have time. Thanks.

**Review by ktoso — APPROVED** [review]
LGTM

**Pull request merged**